aes/aesni/src/aes192/expand.rs

```diff
@@ -1,91 +1,107 @@
+        )
     }
 }
 
 #[inline(always)]
 pub(super) fn expand(key: &[u8; 24]) -> ([__m128i; 13], [__m128i; 13]) {
-    let mut enc_keys: [__m128i; 13] = unsafe { mem::uninitialized() };
-    let mut dec_keys: [__m128i; 13] = unsafe { mem::uninitialized() };
-
-    let (t1, t3) = unsafe {
-        (
-            _mm_loadu_si128(key.as_ptr() as *const __m128i),
-            // we copy garbage for second half, it will be overwritten, so it's fine
-            _mm_loadu_si128(key.as_ptr().offset(16) as *const __m128i),
-        )
-    };
-    enc_keys[0] = t1;
-    enc_keys[1] = t3;
-
-    let (t1, t3) = expand_round!(t1, t3, 0x01);
-    enc_keys[1] = shuffle!(enc_keys[1], t1, 0);
-    enc_keys[2] = shuffle!(t1, t3, 1);
-    let (t1, t3) = expand_round!(t1, t3, 0x02);
-    enc_keys[3] = t1;
-    enc_keys[4] = t3;
-
-    let (t1, t3) = expand_round!(t1, t3, 0x04);
-    enc_keys[4] = shuffle!(enc_keys[4], t1, 0);
-    enc_keys[5] = shuffle!(t1, t3, 1);
-    let (t1, t3) = expand_round!(t1, t3, 0x08);
-    enc_keys[6] = t1;
-    enc_keys[7] = t3;
-
-    let (t1, t3) = expand_round!(t1, t3, 0x10);
-    enc_keys[7] = shuffle!(enc_keys[7], t1, 0);
-    enc_keys[8] = shuffle!(t1, t3, 1);
-    let (t1, t3) = expand_round!(t1, t3, 0x20);
-    enc_keys[9] = t1;
-    enc_keys[10] = t3;
-
-    let (t1, t3) = expand_round!(t1, t3, 0x40);
-    enc_keys[10] = shuffle!(enc_keys[10], t1, 0);
-    enc_keys[11] = shuffle!(t1, t3, 1);
-    let (t1, _) = expand_round!(t1, t3, 0x80);
-    enc_keys[12] = t1;
-
-    dec_keys[0] = enc_keys[0];
-    for i in 1..12 {
-        dec_keys[i] = unsafe { _mm_aesimc_si128(enc_keys[i]) };
-    }
-    dec_keys[12] = enc_keys[12];
+    unsafe {
+        let mut enc_keys: [__m128i; 13] = mem::uninitialized();
+        let mut dec_keys: [__m128i; 13] = mem::uninitialized();
 
-    (enc_keys, dec_keys)
+        macro_rules! store {
+            ($i:expr, $k:expr) => {
+                _mm_store_si128(enc_keys.as_mut_ptr().offset($i), $k);
+                _mm_store_si128(dec_keys.as_mut_ptr().offset($i),
+                    _mm_aesimc_si128($k));
+            }
+        }
+
+        // we are being extra pedantic here to remove out-of-bound access.
+        // this should be optimized out into movups, movsd sequence
+        // note that unaligned load MUST be used here, even though we read
+        // from the array (compiler missoptimizes aligned load)
+        let (k0, k1l) = {
+            let mut t: [u8; 32] = mem::uninitialized();
+            ptr::write(t.as_mut_ptr() as *mut [u8; 24], *key);
+            (
+                _mm_loadu_si128(t.as_ptr() as *const __m128i),
+                _mm_loadu_si128(t.as_ptr().offset(16) as *const __m128i),
+            )
+        };
+
+        _mm_store_si128(enc_keys.as_mut_ptr(), k0);
+        _mm_store_si128(dec_keys.as_mut_ptr(), k0);
+
+        let (k1_2, k2r) = expand_round!(k0, k1l, 0x01);
+        let k1 = shuffle!(k1l, k1_2, 0);
```

Comment:
I am not sure what `shuffle!` does, but `k1l` is partially uninitialized, which means that `k1` might contain uninitialized bytes, so here one needs to be careful.

Reply:
Here shuffle takes half from one variable and other half from the other, thus producing full round key.

Reply:
Makes sense, the first half of `k1l` is properly initialized.